### PR TITLE
Add level0 to db

### PR DESCRIPTION
**Pull request by kdp-cloud**
Adds level 0 elements to the database.

### CODE REVIEW

File: server/app/model.py [hunk 1]
```diff
@@ -88,9 +89,9 @@ class Vector(Base):
     bsa1_overhang: str = Column(String, nullable=True)
     selection: str = Column(String, nullable=True)
     cloning_technique: str = Column(String, nullable=True)  # DNA or PCR synthesis?
-    is_BsmB1_free: str = Column(String, nullable=True)  # TODO: Might be removed...
```
Comment (MaybeJustJames) [inline server/app/model.py:91]: Put these back in please

File: server/app/model.py [hunk 2]
```diff
@@ -88,9 +89,9 @@ class Vector(Base):
...
+    is_BsmB1_free: str = Column(String, nullable=True)  # Might be removed...
     notes: str = Column(String, nullable=True)
-    REase_digest: str = Column(String, nullable=True)  # TODO: Might be removed...
```
Comment (MaybeJustJames) [inline server/app/model.py:93]: Put the TODOs back in please

File: server/app/schemas.py [hunk 3]
```diff
@@ -62,22 +97,55 @@ class LoginUrl(BaseModel):
 
 
 class Reference(BaseModel):
+    """
+    Base Class for defining a Reference.
+    This links the genbank file to the maker.
+
+    Args:
+        BaseModel (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:105]:
```suggestion
```

File: server/app/schemas.py [hunk 4]
```diff
@@ -62,22 +97,55 @@ class LoginUrl(BaseModel):
...
 
 
 class Qualifier(BaseModel):
+    """
+    Base Class for defining qualifiers.
+    Qualifiers are descriptive components
+    from a genbank file.
+
+    Args:
+        BaseModel (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:124]:
```suggestion
```

File: server/app/schemas.py [hunk 5]
```diff
@@ -62,22 +97,55 @@ class LoginUrl(BaseModel):
...
+    """
+    Base Class for defining Features.
+    The features describe like on a map where certain
+    pieces of a construct can be found.
+
+    Features can contain multiple qualifiers to
+    provide more information
+
+    Args:
+        BaseModel (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:146]:
```suggestion
```

File: server/app/schemas.py [hunk 6]
```diff
@@ -89,38 +157,127 @@ class Config:
 
 
 class Annotation(BaseModel):
+    """
+    Base Class defining an Annotation.
+    Annotations are considered as extra
+    information for a construct.
+
+    Args:
+        BaseModel (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:166]:
```suggestion
```

File: server/app/schemas.py [hunk 7]
```diff
@@ -89,38 +157,127 @@ class Config:
...
-    id: int = 0
+class VectorBase(BaseModel):
+    """
+    Base Class for defining a Vector.
+    The Vector class contains all fields that are
+    shared with all the other Vector classes,
+    except the VectorFromGenbank class.
+
+    Args:
+        BaseModel (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:186]:
```suggestion
    Base Class for defining a Vector.
    This class defines common fields that are
    used whether the vector is stored in the
    database, is being sent from the client, or
    is being sent to the client.
```

File: server/app/schemas.py [hunk 8]
```diff
@@ -89,38 +157,127 @@ class Config:
...
+
+
+class VectorFromGenbank(BaseModel):
+    """
+    Base Class for defining a Vector.
+    This class contains all the vector information
+    that is read from a genbank file.
+
+    Args:
+        BaseModel (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:209]:
```suggestion
```

File: server/app/schemas.py [hunk 9]
```diff
@@ -89,38 +157,127 @@ class Config:
...
+
+
+class Vector(VectorBase):
+    """
+    This class inherits from the VectorBase class.
+    This contains additional information shared by
+    most Vector classes.
+
+    Args:
+        VectorBase (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:225]:
```suggestion
```

File: server/app/schemas.py [hunk 10]
```diff
@@ -89,38 +157,127 @@ class Config:
...
+    users: List[User]
+    level: VectorLevel
     gateway_site: str
     vector_type: str
     date: Optional[datetime]
 
+
+class VectorInDB(Vector):
+    """
+    This class inherits from the Vector class.
```
Comment (MaybeJustJames) [inline server/app/schemas.py:243]:
```suggestion
```

File: server/app/schemas.py [hunk 11]
```diff
@@ -89,38 +157,127 @@ class Config:
...
+    It contains all the necessary information for
+    a Vector object when it is stored in the database.
+
+    Args:
+        Vector (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:248]:
```suggestion
```

File: server/app/schemas.py [hunk 12]
```diff
@@ -89,38 +157,127 @@ class Config:
...
+
     class Config:
+        """Orm mode configuration"""
+
         orm_mode = True
+
+
+class VectorToAdd(VectorBase):
+    """
+    This class inherit from the VectorBase class.
```
Comment (MaybeJustJames) [inline server/app/schemas.py:261]:
```suggestion
```

File: server/app/schemas.py [hunk 13]
```diff
@@ -89,38 +157,127 @@ class Config:
...
+    It contains supplemental information provided
+    when a new vector is posted to the database.
+
+    Args:
+        VectorBase (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:266]:
```suggestion
    It contains supplemental information provided
    when a new vector is posted to the database.
    The client sends a string that must be converted into
    a Python datetime object along with unparsed genbank
    data.
```

File: server/app/schemas.py [hunk 14]
```diff
@@ -89,38 +157,127 @@ class Config:
...
+
+
+class VectorOut(Vector):
+    """
+    This class inherit from the Vector class.
+    It contains all the necessary information for
+    a Vector object when it is queried from the database.
+
+    Args:
+        Vector (_type_): _description_
```
Comment (MaybeJustJames) [inline server/app/schemas.py:280]:
```suggestion
    It contains all the necessary information for
    a Vector object when it is queried from the database.
    Not that the `sequence` is **NOT** sent, but instead
    is replaced by just the `sequence_length`. This is because
    the sequence is not used by the client but the length is
    so we save bytes over the wire.
```

File: server/app/vectors.py [hunk 15]
```diff
@@ -1,33 +1,94 @@
...
-@router.get("/vectors/level0", response_model=List[schemas.Vector])
+def vector_to_world(vector: schemas.VectorInDB) -> schemas.VectorOut:
+    """Returns a vector readable by the UI
+
+    Args:
+        vector (schemas.VectorInDB): Vector as stored in DB
+
+    Returns:
+        schemas.VectorOut: Vector readable by UI
+    """
```
Comment (MaybeJustJames) [inline server/app/vectors.py:23]:
```suggestion
    """Returns a vector in the form sent over the wire:
          - replace the sequence by its length

    Args:
        vector: Vector as stored in DB

    Returns:
        schemas.VectorOut: Vector sent over the wire.
    """
```

File: server/app/vectors.py [hunk 16]
```diff
@@ -1,33 +1,94 @@
...
-) -> List[Vector]:
-    return crud.get_level0_for_user(database=database, user=current_user)
+) -> List[schemas.VectorOut]:
+    """Processes GET request from the UI
 
+    Args:
+        database (Session, optional): Database Session.
+        Defaults to Depends(deps.get_db).
+        current_user (schemas.User, optional): User from UI.
+        Defaults to Depends(deps.get_current_user).
```
Comment (MaybeJustJames) [inline server/app/vectors.py:38]:
```suggestion
    """Returns all of the vectors accessible by this user.
```

File: server/app/vectors.py [hunk 17]
```diff
@@ -1,33 +1,94 @@
...
 
-@router.get("/vectors/backbones", response_model=List[schemas.Vector])
-def get_backbones(
-    database: Session = Depends(deps.get_db),
-    current_user: schemas.User = Depends(deps.get_current_user),
-) -> List[Vector]:
-    return crud.get_backbones_for_user(database=database, user=current_user)
+    Returns:
+        List[schemas.VectorOut]: List of Vectors readable for the UI.
```
Comment (MaybeJustJames) [inline server/app/vectors.py:41]:
```suggestion
```

File: server/app/vectors.py [hunk 18]
```diff
@@ -1,33 +1,94 @@
...
+) -> schemas.VectorOut:
+    """Handles POST requests from the UI
+
+    Args:
+        new_vec (schemas.VectorToAdd):
+        A Vector object when posted from UI.
+        database (Session, optional): Database Session.
+        Defaults to Depends(deps.get_db).
+        current_user (schemas.User, optional): Current User.
+        Defaults to Depends(deps.get_current_user).
```
Comment (MaybeJustJames) [inline server/app/vectors.py:63]:
```suggestion
    """Add a Vector to the database
```

File: client/src/Main.elm [hunk 19]
```diff
@@ -1277,6 +1365,71 @@ getAuthentication auth =
...
+        Authenticated usr ->
+            case vector of
+                Level0Vec _ ->
+                    createVector usr.token vector
+
+                BackboneVec _ ->
+                    createVector usr.token vector
+
+                LevelNVec _ ->
+                    createVector usr.token vector
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1427]:
```suggestion
        Authenticated usr ->
            createVector usr.token vector
```

File: client/src/Molecules.elm [hunk 20]
```diff
@@ -37,50 +44,82 @@ type Bsa1Overhang
...
 -}
 type alias Backbone =
     { name : String
     , location : Int
     , bsa1Overhang : Maybe Bsa1Overhang
     , bsmb1Overhang : Maybe Bsmb1Overhang
-    , sequence : String
+    , sequenceLength : Int
+    , bacterialStrain : Maybe String
+    , responsible : String -- Owner of the Level 0 element = User that adds the vector.
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:78]:
```suggestion
    , responsible : String -- TODO: Owner of the Level 0 element = User that adds the vector.
```

File: client/src/Molecules.elm [hunk 21]
```diff
@@ -120,7 +181,16 @@ initBackbone =
     , location = 0
     , bsa1Overhang = Nothing
     , bsmb1Overhang = Nothing
-    , sequence = ""
+    , sequenceLength = 0
+    , bacterialStrain = Nothing
+    , responsible = "" -- Owner of the Level 0 element = User that adds the vector.
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:186]:
```suggestion
    , responsible = "" -- TODO: Owner of the Level 0 element = User that adds the vector.
```

File: client/src/Molecules.elm [hunk 22]
```diff
@@ -226,46 +389,180 @@ level1Decoder =
...
+          , Encode.string <|
+                showBsa1Overhang <|
+                    Maybe.withDefault InvalidBsa1 backbone.bsa1Overhang
+          )
+        , ( "bsmb1_overhang"
+          , Encode.string <|
+                showBsmb1Overhang <|
+                    Maybe.withDefault InvalidBsmb1 backbone.bsmb1Overhang
+          )
+        , ( "sequence", Encode.string "NNNNNNNNNNNNNNNNN---NNNNNNN" ) -- TODO: We do need the sequence!!!
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:448]: This should come from a genbank file no?

File: client/src/Molecules.elm [hunk 23]
```diff
@@ -226,46 +389,180 @@ level1Decoder =
...
+        , ( "group", Encode.string <| level1.group )
+        , ( "notes", Encode.string <| Maybe.withDefault "" level1.notes )
+        , ( "sequenceLength", Encode.int level1.sequenceLength )
+        , ( "genbank_content", Encode.string <| Maybe.withDefault "" level1.genbankContent )
+        ]
+
+
 parseBsa1Overhang : String -> Maybe Bsa1Overhang
 parseBsa1Overhang str =
     case str of
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:502]:
```suggestion
    case String.replace "_" "" str of
```

This allows you to keep only the "XY" cases and remove all the duplications from "X__Y"

**Review by MaybeJustJames — CHANGES_REQUESTED** [review]
Some minor changes to consider

File: client/src/Molecules.elm (hunk 22 above)
Comment (kdp-cloud) [reply client/src/Molecules.elm:448]: Yes indeed! This is an older comment.
It will be deleted in the next commit.

File: client/src/Molecules.elm (hunk 21 above)
Comment (kdp-cloud) [reply client/src/Molecules.elm:186]: The TODO tags will be inserted again in the next commit.

File: server/app/model.py (hunk 1 above)
Comment (kdp-cloud) [reply server/app/model.py:91]: The TODO tags will be inserted again in the next commit.

File: server/app/schemas.py [hunk 24]
```diff
@@ -12,37 +12,68 @@
...
 
 class Token(BaseModel):
+    """
+    Base Class for defining a Token.
+    The token is also linked to a user.
+    """
+
     access_token: str
     token_type: str
     user: User
```
Comment (MaybeJustJames) [inline server/app/schemas.py:69]: This should probably be `UserBase` (client doesn't need to know the `id`). Don't fiddle with this in this PR, it's in bug #43.

File: server/app/schemas.py [hunk 25]
```diff
@@ -12,37 +12,68 @@
...
 
 
 class Provider(BaseModel):
+    """
+    Base Class for defining a provider.
```
Comment (MaybeJustJames) [inline server/app/schemas.py:74]:
```suggestion
    Base Class for defining an OpenID Connect identity provider.
```

File: server/app/vectors.py [hunk 26]
```diff
@@ -1,33 +1,85 @@
...
+) -> schemas.VectorOut:
+    """Handles POST requests from the UI
 
+    Args:
+        new_vec (schemas.VectorToAdd):
+        A Vector object when posted from UI.
+        database (Session, optional): Database Session.
+        Defaults to Depends(deps.get_db).
+        current_user (schemas.User, optional): Current User.
+        Defaults to Depends(deps.get_current_user).
```
Comment (MaybeJustJames) [inline server/app/vectors.py:54]:
```suggestion
```

File: client/src/Molecules.elm [hunk 27]
```diff
@@ -5,29 +5,36 @@ module Molecules exposing
     , ChangeMol(..)
     , Level0
     , Level1
+    , Vector(..)
     , allOverhangs
     , backboneDecoder
+    , backboneEncoder
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:11]: This should not be exposed (you have `vectorDecoder`)

File: client/src/Molecules.elm [hunk 28]
```diff
@@ -5,29 +5,36 @@ module Molecules exposing
...
     , initBackbone
     , initLevel0
     , initLevel1
     , interpretBackboneChange
     , interpretLevel0Change
     , level0Decoder
+    , level0Encoder
     , level1Decoder
+    , levelNEncoder
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:20]: Can you write a `vectorEncoder` that dispatches to these as necessary please?

File: client/src/Molecules.elm [hunk 29]
```diff
@@ -37,50 +44,82 @@ type Bsa1Overhang
     | E__F
     | E__G
     | F__G
+    | InvalidBsa1
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:47]: Why was this added? You should probably use a `Maybe` or `Result` type to handle errors like this.

File: client/src/Molecules.elm [hunk 30]
```diff
@@ -37,50 +44,82 @@ type Bsa1Overhang
...
-{-| What is this?
+{-| All possible overhangs that are produced by a BsmbI digest of a vector.
 -}
 type Bsmb1Overhang
     = W__X
     | W__Z
     | X__Y
     | X__Z
     | Y__Z
+    | InvalidBsmb1
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:58]: Likewise

File: client/src/Main.elm [hunk 31]
```diff
@@ -1277,6 +1365,63 @@ getAuthentication auth =
...
+        Level0Vec vec ->
+            Http.request
+                { method = "POST"
+                , headers = [ Http.header "Authorization" ("Bearer " ++ token) ]
+                , url = "http://localhost:8000/vectors/"
+                , body = Http.jsonBody (level0Encoder vec)
+                , expect = Http.expectJson Level0Created level0Decoder
+                , timeout = Nothing
+                , tracker = Nothing
+                }
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1390]: You can eliminate some of this unnecessary duplication by writing an `authenticatedPost` function.

Then, with a `vectorEncoder` function from the `Molecules` module the entire body of this function becomes:
```elm
createVector : Auth -> Cmd Msg
createVector auth vector =
    case auth of
        Authenticated user ->
            authenticatedPost user.token "http://localhost:8000/vectors/" VectorCreated vectorEncoder
```
or something like that

File: client/src/Main.elm [hunk 32]
```diff
@@ -1277,6 +1365,63 @@ getAuthentication auth =
...
+
+
+postVectors : Auth -> Vector -> Cmd Msg
+postVectors auth vector =
+    case auth of
+        Authenticated usr ->
+            createVector usr.token vector
+
+        _ ->
+            Cmd.none
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1422]: With the above, this can be removed. `createVector` is a better name than `postVector` since `postVector` "leaks" the underlying abstraction (that we're using the HTTP "POST" verb).

**Review by MaybeJustJames — CHANGES_REQUESTED** [review]
Some more relatively minor changes

File: client/src/Molecules.elm (hunk 29 above)
Comment (kdp-cloud) [reply client/src/Molecules.elm]: I've resolved this in the next commit.

File: client/src/Molecules.elm (hunk 30 above)
Comment (kdp-cloud) [reply client/src/Molecules.elm]: I've resolved this in the next commit.

File: client/src/Molecules.elm (hunk 28 above)
Comment (kdp-cloud) [reply client/src/Molecules.elm]: Taken care of in the latest commit.

File: client/src/Main.elm (hunk 31 above)
Comment (kdp-cloud) [reply client/src/Main.elm]: Taken care of in the latest commit.

**Review by kdp-cloud — COMMENTED** [review]
Approved

**Comment by MaybeJustJames** [comment]
> Approved

Huh?

File: client/src/Molecules.elm (hunk 20 above)
Comment (MaybeJustJames) [reply client/src/Molecules.elm:72]: TODO tag not there

File: client/src/Molecules.elm [hunk 33]
```diff
@@ -49,38 +52,68 @@ type Bsmb1Overhang
...
+{-| Type of donor vector which are used to construct a Level 1 Vector.
 -}
 type alias Level0 =
     { name : String
     , location : Int
     , bsa1Overhang : Bsa1Overhang
-    , sequence : String
+    , sequenceLength : Int
+    , bacterialStrain : Maybe String
+    , responsible : String -- Owner of the Level 0 element = User that adds the vector.
```
Comment (MaybeJustJames) [inline client/src/Molecules.elm:91]: TODO comment please

Comment (kdp-cloud) [reply client/src/Molecules.elm]: See commit [b80c2ab](https://github.com/vibbits/GGW-Elm/pull/39/commits/b80c2ab85e4f7335e4d87d141e068e48496a0ed7)

**Review by MaybeJustJames — APPROVED** [review]
Very well done :)
